packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,328 @@
+ * @param props.disabled Whether or not the field is currently disabled.
+ * @param props.onChange Callback to be executed when the value of the input field changes.
+ * @param props.placeholder The placeholder text shown in the input field before the user begins typing.
+ * @param props.errorMessage The message shown below the input field if the value is invalid.
+ * @param props.state The current state of the TextInput. This can be none, valid, or error.
+ * @param props.value The current value of the input field. If a value is passed to this prop, component will be controlled by consumer.
+ * @param props.setValue Callback used to set the value of the input field.
+ * @param props.className className supplied to the TextInput container.
+ */
+const TextInput = React.forwardRef(
+  (
+    {
+      label,
+      description,
+      optional = false,
+      disabled = false,
+      onChange,
+      placeholder,
+      errorMessage,
+      state,
+      value: controlledValue,
+      setValue: setControlledValue,
+      className,
+    }: TextInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) => {
+    const isControlled = typeof controlledValue === 'string';
+    const [uncontrolledValue, setUncontrolledValue] = useState('');
+    const value = isControlled ? controlledValue : uncontrolledValue;
+    const setValue = isControlled ? setControlledValue : setUncontrolledValue;
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.none) {
+        return '#CCC';
+      } else {
+        return state === State.valid ? '#13AA52' : '#CF4A22';
+      }
+    }
+
+    function getInputPaddingFromState() {
+      if (state === State.valid || state === State.error) {
+        return '30px';
+      } else {
+        return optional ? '60px' : '12px';
+      }
+    }
+
+    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
+      if (onChange != undefined) {
+        state = onChange(e.target.value);
+      }
+
+      if (setValue != undefined) {
+        setValue(e.target.value);
+      }
+    }
+
+    const interactionRing = css`
+      transition: all 150ms ease-in-out;
+      transform: scale(0.9, 0.8);
+      border-radius: 7px;
+      position: absolute;
+      top: -3px;
+      bottom: -3px;
+      left: -3px;
+      right: -3px;
+      pointer-events: none;
+      background-color: ${uiColors.gray.light2};
+    `;
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const textInputStyle = css`
```

Comment:
Is there a reason that we're doing all of this styling inside the component? Should probably be able to do out of scope of the component, ditto elsewhere.

Reply:
I just didn't know exactly where all this belonged, so I left it inside the component for the time being. I'll move all of the styling that isn't dependent on the component in some way to be out of scope.